lib/charms/traefik_route_k8s/v0/traefik_route.py

```diff
@@ -0,0 +1,174 @@
+    )
+    self.framework.observe(
+        self.traefik_route.on.response, self.ingress_per_unit.respond
+    )
+```
+"""
+import logging
+from typing import Optional
+
+import yaml
+from ops.charm import CharmBase, RelationEvent, CharmEvents
+from ops.framework import EventSource, Object
+from ops.model import Relation
+
+# The unique Charmhub library identifier, never change it
+LIBID = ""
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 0
+
+log = logging.getLogger(__name__)
+
+
+class TraefikRouteException(RuntimeError):
+    """Base class for exceptions raised by TraefikRoute."""
+
+
+class UnauthorizedError(TraefikRouteException):
+    """Raised when the unit needs leadership to perform some action."""
+
+
+class TraefikRouteProviderReadyEvent(RelationEvent):
+    """Event emitted when Traefik is ready to provide ingress for a routed unit."""
+
+
+class TraefikRouteRequirerReadyEvent(RelationEvent):
+    """Event emitted when a unit requesting ingress has provided all data Traefik needs."""
+
+
+class TraefikRouteRequirerEvents(CharmEvents):
+    """Container for TraefikRouteRequirer events."""
+    ready = EventSource(TraefikRouteRequirerReadyEvent)
+
+
+class TraefikRouteProviderEvents(CharmEvents):
+    """Container for TraefikRouteProvider events."""
+    ready = EventSource(TraefikRouteProviderReadyEvent)
+
+
+class TraefikRouteProvider(Object):
+    """Implementation of the provider of traefik_route.
+
+    This will presumably be owned by a Traefik charm.
+    The main idea is that Traefik will observe the `ready` event and, upon
+    receiving it, will fetch the config from the TraefikRoute's application databag,
+    apply it, and update its own app databag to let Route know that the ingress
+    is there.
+    """
+    on = TraefikRouteProviderEvents()
+
+    def __init__(self, charm: CharmBase, relation_name: str = 'traefik-route'):
+        """Constructor for TraefikRouteProvider.
+
+        Args:
+            charm: The charm that is instantiating the instance.
+            relation_name: The name of the relation relation_name to bind to
+                (defaults to "traefik-route").
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.framework.observe(self.charm.on[relation_name].relation_changed,
+                               self._on_relation_changed)
+
+    def _on_relation_changed(self, event: RelationEvent):
+        if self.is_ready(event.relation):
+            # todo check data is valid here?
+            self.on.ready.emit(event.relation)
+
+    @staticmethod
+    def is_ready(relation: Relation) -> bool:
+        """Whether TraefikRoute is ready on this relation: i.e. the remote app shared the config."""
+        return 'config' in relation.data[relation.app]
+
+    @staticmethod
```

Comment:
Do these need to be static?

Reply:
No, but my pycharm linter complains if they aren't. Do they need to not be static?

Reply:
Just seems odd to me, but I guess it's not a huge problem.